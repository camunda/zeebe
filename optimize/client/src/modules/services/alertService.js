/*
 * Copyright Camunda Services GmbH and/or licensed to Camunda Services GmbH
 * under one or more contributor license agreements. Licensed under a proprietary license.
 * See the License.txt file for more information. You may not use this file
 * except in compliance with the proprietary license.
 */

import {del, get, post, put} from 'request';
import {track} from 'tracking';

export async function loadAlerts(collection) {
  const response = await get(`api/collection/${collection}/alerts`);
  return await response.json();
}

export async function addAlert(alert) {
  const response = await post(`api/alert`, alert);
  const json = await response.json();
  track('createAlert', {entityId: json.id});
  return response;
}

export async function editAlert(id, alert) {
  const response = await put(`api/alert/${id}`, alert);
  track('updateAlert', {entityId: id});
  return response;
}

export async function removeAlert(id) {
  const response = await del(`api/alert/${id}`);
  track('deleteAlert', {entityId: id});
  return response;
}
