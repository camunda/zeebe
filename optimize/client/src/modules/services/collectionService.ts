/*
 * Copyright Camunda Services GmbH and/or licensed to Camunda Services GmbH
 * under one or more contributor license agreements. Licensed under a proprietary license.
 * See the License.txt file for more information. You may not use this file
 * except in compliance with the proprietary license.
 */

import {Source} from 'types';
import {put} from 'request';

export async function addSources(collectionId: string, sources: Source[]) {
  return await put(`api/collection/${collectionId}/scope`, sources);
}

export function getCollection(path: string) {
  const collectionMatch = /\/collection\/([^/]+)/g.exec(path);
  return collectionMatch && collectionMatch[1];
}
