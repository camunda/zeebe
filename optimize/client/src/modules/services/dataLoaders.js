/*
 * Copyright Camunda Services GmbH and/or licensed to Camunda Services GmbH
 * under one or more contributor license agreements. Licensed under a proprietary license.
 * See the License.txt file for more information. You may not use this file
 * except in compliance with the proprietary license.
 */

import {get, post} from 'request';

export {
  loadProcessDefinitionXml,
  loadDecisionDefinitionXml,
  loadVariables,
  loadInputVariables,
  loadOutputVariables,
} from './dataLoaders.ts';

export async function getFlowNodeNames(processDefinitionKey, processDefinitionVersion, tenantId) {
  if (processDefinitionKey && processDefinitionVersion) {
    const payload = {
      processDefinitionKey,
      processDefinitionVersion,
    };

    if (tenantId) {
      payload.tenantId = tenantId;
    }

    const response = await post(`api/flow-node/flowNodeNames`, payload);

    const json = await response.json();

    return await json.flowNodeNames;
  } else {
    return {};
  }
}

export async function checkDeleteConflict(id, entity) {
  const response = await get(`api/${entity}/${id}/delete-conflicts`);
  return await response.json();
}
