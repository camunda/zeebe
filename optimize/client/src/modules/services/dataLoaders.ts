/*
 * Copyright Camunda Services GmbH and/or licensed to Camunda Services GmbH
 * under one or more contributor license agreements. Licensed under a proprietary license.
 * See the License.txt file for more information. You may not use this file
 * except in compliance with the proprietary license.
 */

import {ProcessFilter, Variable} from 'types';
import {get, post} from 'request';

export async function loadProcessDefinitionXml(
  key?: string,
  version?: string,
  tenantId?: string | null
): Promise<string | null> {
  const payload: {key?: string; version?: string; tenantId?: string} = {key, version};
  if (tenantId) {
    payload.tenantId = tenantId;
  }

  try {
    const response = await get('api/definition/process/xml', payload);

    return await response.text();
  } catch (e) {
    return null;
  }
}

export async function loadDecisionDefinitionXml(
  key: string,
  version?: string,
  tenantId?: string | null
): Promise<string | null> {
  const payload: {key: string; version?: string; tenantId?: string} = {key, version};
  if (tenantId) {
    payload.tenantId = tenantId;
  }
  try {
    const response = await get('api/definition/decision/xml', payload);

    return await response.text();
  } catch (e) {
    return null;
  }
}

const loadVariablesFrom =
  <P>(endpoint: string) =>
  async (payload: P) => {
    const response = await post(endpoint, payload);

    return (await response.json()) as Variable[];
  };

type Process = {
  processDefinitionKey: string;
  processDefinitionVersions: string[];
  tenantIds: (string | null)[];
};

type LoadVariablesPayload = {
  processesToQuery: Process[];
  filter: ProcessFilter[];
};
export const loadVariables = loadVariablesFrom<LoadVariablesPayload>('api/variables');

type LoadDecisionVariablesPayload = {
  decisionDefinitionKey: string;
  decisionDefinitionVersions: string[];
  tenantIds: (string | null)[];
};
export const loadInputVariables = loadVariablesFrom<LoadDecisionVariablesPayload>(
  'api/decision-variables/inputs/names'
);
export const loadOutputVariables = loadVariablesFrom<LoadDecisionVariablesPayload>(
  'api/decision-variables/outputs/names'
);
