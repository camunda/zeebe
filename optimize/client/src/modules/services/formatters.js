/*
 * Copyright Camunda Services GmbH and/or licensed to Camunda Services GmbH
 * under one or more contributor license agreements. Licensed under a proprietary license.
 * See the License.txt file for more information. You may not use this file
 * except in compliance with the proprietary license.
 */

import {parseISO, isValid} from 'date-fns';

import {format} from 'dates';
import {t} from 'translation';

import {formatTenantName} from './formatters.tsx';

export {createDurationFormattingOptions, formatFileName, formatTenantName} from './formatters.tsx';
export {UNAUTHORIZED_TENANT_ID} from './tenantService.ts';

const scaleUnits = [
  {exponent: 18, label: 'quintillion'},
  {exponent: 15, label: 'quadrillion'},
  {exponent: 12, label: 'trillion'},
  {exponent: 9, label: 'billion'},
  {exponent: 6, label: 'million'},
  {exponent: 3, label: 'thousand'},
];

function getNumberOfDigits(x) {
  // https://stackoverflow.com/a/28203456
  return Math.max(Math.floor(Math.log10(Math.abs(x))), 0) + 1;
}

export function frequency(number, precision) {
  if (!number && number !== 0) {
    return '--';
  }

  const intl = new Intl.NumberFormat();

  if (precision) {
    const digitsFactor = 10 ** getNumberOfDigits(number);
    const precisionFactor = 10 ** precision;
    const roundedToPrecision =
      (Math.round((number / digitsFactor) * precisionFactor) / precisionFactor) * digitsFactor;

    for (let i = 0; i < scaleUnits.length; i++) {
      const {exponent, label} = scaleUnits[i];
      if (Math.abs(roundedToPrecision) >= 10 ** exponent) {
        const shortened = roundedToPrecision / 10 ** exponent;
        return (
          intl.format(shortened) +
          ' ' +
          t(`common.unit.${label}.label${shortened !== 1 ? '-plural' : ''}`)
        );
      }
    }
    return intl.format(roundedToPrecision);
  }
  return intl.format(number);
}

export function percentage(number) {
  if (!number && number !== 0) {
    return '--';
  }

  return Number(Number(number).toFixed(2)) + '%';
}

export function getRelativeValue(data, total) {
  if (!data && data !== 0) {
    return '--';
  }
  if (data === 0) {
    return '0%';
  }
  return Math.round((data / total) * 1000) / 10 + '%';
}

export function camelCaseToLabel(type) {
  return type.replace(/([A-Z])/g, ' $1').replace(/^./, (str) => str.toUpperCase());
}

export function objectifyResult(result) {
  return result.reduce((acc, {key, value}) => {
    acc[key] = value;
    return acc;
  }, {});
}

export function formatReportResult(data, result) {
  const groupBy = data.groupBy;

  let unit;
  if (groupBy.value && groupBy.type.includes('Date')) {
    unit = determineUnit(groupBy.value.unit, result);
  } else if (
    groupBy.value &&
    groupBy.type.toLowerCase().includes('variable') &&
    groupBy.value.type === 'Date'
  ) {
    unit = determineUnit(data.configuration.groupByDateVariableUnit, result);
  }

  if (!unit || !result) {
    // the result data is no time series
    return [...result];
  }

  let dateFormat = getDateFormat(unit);
  // The added space is to make sure all dates are string,
  // since integer keys in objects does not preserve order
  if (unit === 'year') {
    dateFormat += ' ';
  }

  const formattedResult = result.map((entry) => {
    const date = parseISO(entry.label);
    if (!isValid(date)) {
      return entry;
    }
    return {
      ...entry,
      label: format(date, dateFormat),
    };
  });

  return formattedResult;
}

function determineUnit(unit, resultData) {
  if (unit === 'automatic') {
    return determineUnitForAutomaticIntervalSelection(resultData);
  } else {
    // in this case the unit was already defined by the user
    // and can just directly be used.
    return unit;
  }
}

function determineUnitForAutomaticIntervalSelection(resultData) {
  if (resultData.length > 1) {
    const firstEntry = parseISO(resultData[0].key);
    const secondEntry = parseISO(resultData[1].key);
    const intervalInMs = Math.abs(firstEntry - secondEntry);

    const intervals = [
      {value: 1000 * 60 * 60 * 24 * 30 * 12, unit: 'year'},
      {value: 1000 * 60 * 60 * 24 * 30, unit: 'month'},
      {value: 1000 * 60 * 60 * 24, unit: 'day'},
      {value: 1000 * 60 * 60, unit: 'hour'},
      {value: 0, unit: 'second'},
      {value: -Infinity, unit: 'day'},
    ];
    return intervals.find(({value}) => intervalInMs >= value).unit;
  } else {
    return 'day';
  }
}

function getDateFormat(unit) {
  let dateFormat;
  switch (unit) {
    case 'hour':
      dateFormat = 'yyyy-MM-dd HH:00:00';
      break;
    case 'day':
    case 'week':
      dateFormat = 'yyyy-MM-dd';
      break;
    case 'month':
      dateFormat = 'MMM yyyy';
      break;
    case 'year':
      dateFormat = 'yyyy';
      break;
    case 'second':
    default:
      dateFormat = 'yyyy-MM-dd HH:mm:ss';
  }
  return dateFormat;
}

export function formatVersions(versions) {
  if (versions.length === 1 && versions[0] === 'all') {
    return t('common.all');
  } else if (versions.length === 1 && versions[0] === 'latest') {
    return t('common.definitionSelection.latest');
  } else if (versions.length) {
    return versions.join(', ');
  }

  return t('common.none');
}

export function formatTenants(tenantIds, tenantInfo, showOnlyTenant) {
  if (showOnlyTenant) {
    return formatTenantName(tenantInfo[0]);
  }

  if (tenantIds.length === 0) {
    return t('common.none');
  }

  if (tenantInfo && tenantInfo.length > 1) {
    if (tenantInfo.length === tenantIds.length) {
      return t('common.all');
    } else {
      return tenantIds
        .map((tenantId) =>
          formatTenantName(tenantInfo.find(({id}) => id === tenantId) ?? {id: tenantId})
        )
        .join(', ');
    }
  }

  return '';
}

export function formatLabel(label, formatNumbersOnly) {
  const MAX_LENGTH = 50;

  if (!label || typeof label === 'object' || label.toString().length <= MAX_LENGTH) {
    return label;
  }

  // too long string
  if (isNaN(label)) {
    return formatNumbersOnly ? label : label.slice(0, MAX_LENGTH) + '...';
  }

  // too long number
  return Number.parseFloat(label).toExponential();
}

export {
  duration,
  getHighlightedText,
  convertDurationToObject,
  convertDurationToSingleNumber,
  convertToDecimalTimeUnit,
  convertToMilliseconds,
} from './formatters.tsx';
