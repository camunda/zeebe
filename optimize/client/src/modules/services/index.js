/*
 * Copyright Camunda Services GmbH and/or licensed to Camunda Services GmbH
 * under one or more contributor license agreements. Licensed under a proprietary license.
 * See the License.txt file for more information. You may not use this file
 * except in compliance with the proprietary license.
 */

export {
  getFlowNodeNames,
  checkDeleteConflict,
  loadDecisionDefinitionXml,
  loadInputVariables,
  loadOutputVariables,
  loadProcessDefinitionXml,
  loadVariables,
} from './dataLoaders';

export {numberParser} from './NumberParser';
export {incompatibleFilters} from './incompatibleFilters.ts';
export {default as reportConfig, createReportUpdate, getDefaultSorting} from './reportConfig';
export {getDiagramElementsBetween} from './diagramServices';
export {default as getTooltipText} from './getTooltipText';
export {default as getScreenBounds} from './getScreenBounds';
export {loadEntity, loadReports, updateEntity, deleteEntity, copyReport} from './entityService';
export {UNAUTHORIZED_TENANT_ID} from './tenantService.ts';
export {loadEntities, copyEntity, createEntity, getEntityIcon} from './entityService.tsx';

export {
  evaluateReport,
  isDurationReport,
  loadRawData,
  getReportResult,
  processResult,
  isAlertCompatibleReport,
  isCategoricalBar,
  isCategorical,
  TEXT_REPORT_MAX_CHARACTERS,
  isTextTileTooLong,
  isTextTileValid,
} from './reportService';

export {addSources, getCollection} from './collectionService';

export * as formatters from './formatters';

export function getRandomId() {
  return Math.random().toString(36).slice(2);
}

export function capitalize(string) {
  return string.charAt(0).toUpperCase() + string.slice(1);
}

export {loadAlerts, addAlert, removeAlert, editAlert} from './alertService';
export {default as ignoreFragments} from './ignoreFragments';
export {default as isReactElement} from './isReactElement';
export {loadDefinitions} from './loadDefinitions';
