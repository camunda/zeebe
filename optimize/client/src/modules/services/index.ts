/*
 * Copyright Camunda Services GmbH and/or licensed to Camunda Services GmbH
 * under one or more contributor license agreements. Licensed under a proprietary license.
 * See the License.txt file for more information. You may not use this file
 * except in compliance with the proprietary license.
 */

export function getRandomId() {
  return Math.random().toString(36).slice(2);
}

export {loadReports, loadEntities, copyEntity, createEntity, getEntityIcon} from './entityService';
export {UNAUTHORIZED_TENANT_ID} from './tenantService';
export * as formatters from './formatters';
export {
  loadProcessDefinitionXml,
  loadDecisionDefinitionXml,
  loadInputVariables,
  loadOutputVariables,
  loadVariables,
} from './dataLoaders';
export {numberParser} from './NumberParser';
export {
  TEXT_REPORT_MAX_CHARACTERS,
  isTextTileTooLong,
  isTextTileValid,
  loadRawData,
  evaluateReport,
} from './reportService';
export {addSources, getCollection} from './collectionService';
export {default as getScreenBounds} from './getScreenBounds';
export {default as ignoreFragments} from './ignoreFragments';
export {default as isReactElement} from './isReactElement';
export {incompatibleFilters} from './incompatibleFilters';
export {loadDefinitions} from './loadDefinitions';

export type {Definition} from './loadDefinitions';
export type {ReportEvaluationPayload} from './reportService';
