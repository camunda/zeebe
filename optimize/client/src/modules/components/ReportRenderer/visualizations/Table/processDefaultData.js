/*
 * Copyright Camunda Services GmbH and/or licensed to Camunda Services GmbH
 * under one or more contributor license agreements. Licensed under a proprietary license.
 * See the License.txt file for more information. You may not use this file
 * except in compliance with the proprietary license.
 */

import {reportConfig, formatters, processResult} from 'services';
import {t} from 'translation';
import {formatValue} from '../../service';

import {formatLabelsForTableBody, sortColumns} from './service';

const {formatReportResult, getRelativeValue, frequency, duration} = formatters;

export default function processDefaultData({report}, processVariables = []) {
  const {data, result, reportType} = report;
  const {configuration, view, groupBy} = data;
  const {
    hideAbsoluteValue,
    hideRelativeValue,
    tableColumns: {columnOrder},
    precision,
  } = configuration;

  const groupedByDuration = groupBy.type === 'duration';
  const instanceCount = result.instanceCount || 0;
  const config = reportConfig[reportType];

  const isMultiMeasure = result.measures.length > 1;

  const viewString = config.view.find(({matcher}) => matcher(data)).label();
  const groupString = config.group.find(({matcher}) => matcher(data)).label();

  const head = [];
  let body = [];

  if (reportType === 'process' && (groupBy.type === 'duration' || groupBy.type.includes('Date'))) {
    head.push(viewString + ' ' + groupString);
  } else if (view.entity === 'processInstance' && groupBy.type === 'variable') {
    head.push(
      `${viewString} ${t('report.table.rawData.variable')}: ${
        getVariableLabel(processVariables, groupBy.value) || groupBy.value.name
      }`
    );
  } else if (['inputVariable', 'outputVariable'].includes(groupBy.type)) {
    head.push(`${t('report.groupBy.' + groupBy.type)}: ${groupBy.value.name}`);
  } else if (view.entity === 'incident' && groupBy.type === 'flowNodes') {
    head.push(t('common.incident.byFlowNode'));
  } else {
    head.push(groupString);
  }

  result.measures.forEach((measure) => {
    const result = processResult({...report, result: measure});
    const formattedResult = formatReportResult(data, result.data);
    if (body.length === 0) {
      formattedResult.forEach(({label, key}) => {
        body.push([groupedByDuration ? duration(label, precision) : label || key]);
      });
    }

    if (measure.property === 'frequency') {
      if (!hideAbsoluteValue) {
        const title = t('report.view.count');
        head.push({label: title, id: title, sortable: !isMultiMeasure});
        formattedResult.forEach(({value}, idx) => {
          body[idx].push(frequency(value, precision));
        });
      }
      if (!hideRelativeValue) {
        const title = t('report.table.relativeFrequency');
        head.push({label: title, id: title, sortable: !isMultiMeasure});
        formattedResult.forEach(({value}, idx) => {
          body[idx].push(getRelativeValue(value, instanceCount));
        });
      }
    } else if (measure.property === 'duration') {
      const title = `${
        measure.userTaskDurationTime
          ? `${t('report.config.userTaskDuration.' + measure.userTaskDurationTime)} `
          : ''
      }${
        view.entity === 'incident' ? t('report.view.resolutionDuration') : t('report.view.duration')
      } - ${t('report.config.aggregationShort.' + measure.aggregationType.type, {
        value: formatValue(measure.aggregationType.value, measure.property, precision),
      })}`;

      head.push({label: title, id: title, sortable: !isMultiMeasure});
      formattedResult.forEach(({value}, idx) => {
        body[idx].push(duration(value, precision));
      });
    }
  });

  const {sortedHead, sortedBody} = sortColumns(head, body, columnOrder);

  body = formatLabelsForTableBody(sortedBody);

  return {head: sortedHead, body};
}

function getVariableLabel(variables, {name, type}) {
  return variables.find((variable) => variable.name === name && variable.type === type)?.label;
}
