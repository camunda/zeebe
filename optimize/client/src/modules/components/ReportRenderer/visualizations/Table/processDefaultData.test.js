/*
 * Copyright Camunda Services GmbH and/or licensed to Camunda Services GmbH
 * under one or more contributor license agreements. Licensed under a proprietary license.
 * See the License.txt file for more information. You may not use this file
 * except in compliance with the proprietary license.
 */

import update from 'immutability-helper';

import {formatters} from 'services';

import processDefaultData from './processDefaultData';

const {getRelativeValue} = formatters;

jest.mock('services', () => ({
  ...jest.requireActual('services'),
  formatters: {
    getRelativeValue: jest.fn(),
    formatReportResult: jest.fn().mockReturnValue([
      {key: 'a', value: 1, label: 'a name'},
      {key: 'b', value: 2, label: 'b'},
      {key: 'c', value: 3, label: 'c'},
    ]),
    duration: (a) => 'Duration: ' + a,
    frequency: (a) => a,
  },
}));

const report = {
  reportType: 'process',
  combined: false,
  data: {
    groupBy: {type: 'startDate', value: {unit: 'automatic'}},
    view: {entity: 'processInstance', properties: ['frequency']},
    configuration: {
      tableColumns: {
        includeNewVariables: true,
        includedColumns: [],
        excludedColumns: [],
        columnOrder: [],
      },
      hideRelativeValue: true,
      hideAbsoluteValue: false,
    },
    visualization: 'table',
  },
  result: {
    measures: [{property: 'frequency', data: []}],
    instanceCount: 5,
  },
};

const props = {
  report,
};

it('should display data for key-value pairs', async () => {
  expect(processDefaultData(props).body).toEqual([
    ['a name', 1],
    ['b', 2],
    ['c', 3],
  ]);
});

it('should format the label when the report is grouped by duration', () => {
  const newProps = {
    ...props,
    report: update(report, {data: {groupBy: {$set: {type: 'duration'}}}}),
  };
  expect(processDefaultData(newProps).body).toEqual([
    ['Duration: a name', 1],
    ['Duration: b', 2],
    ['Duration: c', 3],
  ]);
});

const newProps = {
  ...props,
  report: {
    ...report,
    data: {
      ...report.data,
      configuration: {
        hideAbsoluteValue: true,
        hideRelativeValue: false,
        tableColumns: {
          columnOrder: [],
        },
      },
    },
  },
};

it('should not include absolute values if if it is hidden in the configuration', () => {
  getRelativeValue.mockClear();
  getRelativeValue.mockReturnValue('12.3%');

  expect(processDefaultData(newProps).body).toEqual([
    ['a name', '12.3%'],
    ['b', '12.3%'],
    ['c', '12.3%'],
  ]);
});

it('should display the relative percentage for frequency views', () => {
  getRelativeValue.mockClear();
  getRelativeValue.mockReturnValue('12.3%');

  const tableData = processDefaultData(newProps);

  expect(getRelativeValue).toHaveBeenCalledWith(1, 5);
  expect(getRelativeValue).toHaveBeenCalledWith(2, 5);
  expect(getRelativeValue).toHaveBeenCalledWith(3, 5);

  expect(tableData.body[0][1]).toBe('12.3%');
});

it('should display the relative percentage for frequency views for DMN', () => {
  getRelativeValue.mockClear();
  getRelativeValue.mockReturnValue('12.3%');

  const dmnProps = {
    ...newProps,
    report: {
      ...newProps.report,
      result: {
        ...newProps.report.result,
        instanceCount: 18,
      },
    },
  };

  const dmnTableData = processDefaultData(dmnProps);

  expect(getRelativeValue).toHaveBeenCalledWith(1, 18);
  expect(getRelativeValue).toHaveBeenCalledWith(2, 18);
  expect(getRelativeValue).toHaveBeenCalledWith(3, 18);

  expect(dmnTableData.body[0][1]).toBe('12.3%');
});

it('should display multi-measure reports', () => {
  const table = processDefaultData({
    report: update(report, {
      data: {view: {properties: {$set: ['frequency', 'duration']}}},
      result: {
        measures: {
          $set: [
            {property: 'frequency', data: []},
            {property: 'duration', aggregationType: {type: 'avg', value: null}, data: []},
          ],
        },
      },
    }),
  });

  expect(table).toMatchSnapshot();
});

it('should handle undefined processVariables for variable reports', () => {
  const result = processDefaultData({
    report: {
      ...props.report,
      data: {
        ...props.report.data,
        groupBy: {type: 'variable', value: {name: 'variable1', type: 'string'}},
      },
    },
  });

  expect(result).toEqual({
    body: [
      ['a name', 1],
      ['b', 2],
      ['c', 3],
    ],
    head: ['Process instance Var: variable1', {id: 'Count', label: 'Count', sortable: true}],
  });
});

it('should handle processVariables with a matching variable', () => {
  const variableReportProps = {
    report: {
      ...props.report,
      data: {
        ...props.report.data,
        groupBy: {type: 'variable', value: {name: 'variable1', type: 'string'}},
      },
    },
  };

  const result = processDefaultData(variableReportProps, [
    {name: 'variable1', type: 'string', label: 'Variable One'},
    {name: 'variable2', type: 'number', label: 'Variable Two'},
  ]);

  expect(result).toEqual({
    body: [
      ['a name', 1],
      ['b', 2],
      ['c', 3],
    ],
    head: ['Process instance Var: Variable One', {id: 'Count', label: 'Count', sortable: true}],
  });
});
