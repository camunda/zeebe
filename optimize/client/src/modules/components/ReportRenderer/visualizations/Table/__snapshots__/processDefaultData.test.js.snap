// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should display multi-measure reports 1`] = `
Object {
  "body": Array [
    Array [
      "a name",
      1,
      "Duration: 1",
    ],
    Array [
      "b",
      2,
      "Duration: 2",
    ],
    Array [
      "c",
      3,
      "Duration: 3",
    ],
  ],
  "head": Array [
    "Process instance Start date",
    Object {
      "id": "Count",
      "label": "Count",
      "sortable": false,
    },
    Object {
      "id": "Duration - Avg",
      "label": "Duration - Avg",
      "sortable": false,
    },
  ],
}
`;
